### PR TITLE
improve bgp update timer to cover bgp session down case

### PR DESCRIPTION
**Pull request by jcaiMR**
### Description of PR

Summary:
test_bgp_update_timer.py used to cover SONiC sends BGP updates to neighbors after received updates from other neighbors within 1s. Currently it missing the case when bgp session down with neighbors, SONiC can send update to other neighbors in time.
So add this to improve test case coverage. 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [x] 201911
- [x] 202012
- [x] 202205

#### What is the motivation for this PR?
Improve test case coverage as mentioned in description. 

#### How did you do it?
Run cases on T0 and T1

#### How did you verify/test it?
image

### CODE REVIEW

File: tests/bgp/test_bgp_update_timer.py [hunk 1]
```diff
@@ -305,3 +310,58 @@ def is_neighbor_sessions_established(duthost, neighbors):
...
+                pytest.fail("announce route %s from n0 to dut failed" % route["prefix"])
+        # close bgp session n0, monitor withdraw info from dut to n1
+        bgp_pcap = BGP_DOWN_LOG_TMPL
+        with log_bgp_updates(duthost, "any", bgp_pcap, n0.namespace):
+            duthost.shell("config bgp shutdown neighbor {}".format(n0.name))
+            global current_time 
+            current_time = time.time()        
+
+        with tempfile.NamedTemporaryFile() as tmp_pcap:
+                duthost.fetch(src=bgp_pcap, dest=tmp_pcap.name, flat=True)
```
Comment (StormLiangMS) [inline tests/bgp/test_bgp_update_timer.py:346]: @jcaiMR indent issue, use 4 spaces.

File: tests/bgp/test_bgp_update_timer.py [hunk 2]
```diff
@@ -305,3 +310,58 @@ def is_neighbor_sessions_established(duthost, neighbors):
...
+                duthost.file(path=bgp_pcap, state="absent")
+                bgp_updates = bgp_update_packets(tmp_pcap.name)
+                
+        for bgp_update in bgp_updates:
+            for i, route in enumerate(constants.routes):
+                if match_bgp_update(bgp_update, n1.peer_ip, n1.ip, "withdraw", route):
+                    withdraw_intervals[i] = bgp_update.time - current_time
```
Comment (StormLiangMS) [inline tests/bgp/test_bgp_update_timer.py:353]: There are 2 loops, this one works, but the efficiency is low, we may revisit this later with better algo. @jcaiMR

File: tests/bgp/test_bgp_update_timer.py [hunk 3]
```diff
@@ -305,3 +310,58 @@ def is_neighbor_sessions_established(duthost, neighbors):
...
+        
+        for i, route in enumerate(constants.routes):
+            if withdraw_intervals[i] >= constants.update_interval_threshold:
+                pytest.fail("withdraw updates interval %d exceeds threshold %d" % (constants.update_interval_threshold))
```
Comment (StormLiangMS) [inline tests/bgp/test_bgp_update_timer.py:357]: There are two %d, but only one variable, pls use black or other tools to check the format.  @jcaiMR

File: tests/bgp/test_bgp_update_timer.py (hunk 1 above)
Comment (jcaiMR) [reply tests/bgp/test_bgp_update_timer.py:346]: Will apply in latest diff.

**Review by StormLiangMS — APPROVED** [review]
approved with a suggestion. 
should we consider to increase the route number, here it is only 6, in production, it should be around thousand.

File: tests/bgp/test_bgp_update_timer.py [hunk 4]
```diff
@@ -305,3 +362,72 @@ def is_neighbor_sessions_established(duthost, neighbors):
...
+            n0.announce_route(route)
+            time.sleep(constants.sleep_interval)
+            dut_route = duthost.get_route(route["prefix"], n0.namespace)
+            if not dut_route:
+                pytest.fail("announce route %s from n0 to dut failed" % route["prefix"])
+        # close bgp session n0, monitor withdraw info from dut to n1
+        bgp_pcap = BGP_DOWN_LOG_TMPL
+        with log_bgp_updates(duthost, "any", bgp_pcap, n0.namespace):
+            duthost.shell("config bgp shutdown neighbor {}".format(n0.name))
+            global current_time
```
Comment (lolyu) [inline tests/bgp/test_bgp_update_timer.py:402]: the global keyword is useless here.

**Review by lolyu — APPROVED** [review]

**Review by StormLiangMS — APPROVED** [review]

**Pull request merged**